Zend/zend_compile.c

```diff
@@ -5711,6 +5717,93 @@ static zend_bool zend_is_valid_default_value(zend_type type, zval *value)
 	return 0;
 }
 
+static void zend_compile_attribute(zval *v, zend_ast *ast) /* {{{ */
+{
+	ZEND_ASSERT(ast->kind == ZEND_AST_ATTRIBUTE);
+
+	array_init_size(v, 1 + (ast->child[1] ? zend_ast_get_list(ast->child[1])->children : 0));
+	add_next_index_str(v, zend_resolve_class_name(zend_ast_get_str(ast->child[0]), ZEND_NAME_NOT_FQ));
+
+	if (ast->child[1]) {
+		zend_ast_list *list;
+		uint32_t i;
+		zval tmp;
+
+		ZEND_ASSERT(ast->child[1]->kind == ZEND_AST_ARG_LIST);
+
+		ZVAL_NULL(&tmp);
+
+		for (list = zend_ast_get_list(ast->child[1]), i = 0; i < list->children; i++) {
+			zend_const_expr_to_zval(zend_hash_next_index_insert(Z_ARRVAL_P(v), &tmp), list->child[i]);
+		}
+	}
+}
+/* }}} */
+
+static HashTable *zend_compile_attributes(zend_ast *ast, int target) /* {{{ */
+{
+	HashTable *attr;
+
+	zend_ast_list *list;
+	uint32_t i;
+
+	zval tmp;
+	zend_attributes_internal_validator validator = NULL;
+
+	ZVAL_NULL(&tmp);
+
+	ZEND_ASSERT(ast->kind == ZEND_AST_ATTRIBUTE_LIST);
+
+	ALLOC_HASHTABLE(attr);
+	zend_hash_init(attr, zend_ast_get_list(ast)->children, NULL, ZVAL_PTR_DTOR, 0);
+
+	for (list = zend_ast_get_list(ast), i = 0; i < list->children; i++) {
+		zend_ast *el = list->child[i];
+		zend_string *name;
+
+		zval a;
+		zval *x;
+
+		zend_compile_attribute(&a, el);
+
+		name = zend_string_tolower(Z_STR_P(zend_hash_index_find(Z_ARRVAL(a), 0)));
+		x = zend_hash_find(attr, name);
+
+		// validate internal attribute
+		validator = (zend_attributes_internal_validator)zend_hash_find_ptr(&zend_attributes_internal_validators, name);
+
+		if (validator != NULL) {
+			validator(&a, target);
+		}
+
+		if (x) {
+			ZEND_ASSERT(Z_TYPE_P(x) == IS_ARRAY);
+
+			if (Z_TYPE_P(zend_hash_index_find(Z_ARRVAL_P(x), 0)) == IS_ARRAY) {
+				add_next_index_zval(x, &a);
+			} else {
+				zval array;
+
+				ZEND_ASSERT(Z_TYPE_P(zend_hash_index_find(Z_ARRVAL_P(x), 0)) == IS_STRING);
+
+				Z_ADDREF_P(x);
+
+				array_init(&array);
+				add_next_index_zval(&array, x);
+				add_next_index_zval(&array, &a);
+				zend_hash_update(attr, name, &array);
+			}
+		} else {
+			zend_hash_add(attr, name, &a);
+		}
```

Comment:
So the representation here is either `lcname => [name, args...]` or `lcname => [[name, args...], [name, args...], ...]`?

Do we actually need an efficient lookup by lcname for anything? If not, I would just always store the attributes as a linear array.

Reply:
Storing them in a linear array would be a lot simpler and performance would still be fine (especially if we store lcname within attribute data / struct). I took advantage of attributes using string keys to store parameter attributes in the same array using numeric indexes. If this is changed we need a new way to identify these. If attributes are stored as structs we could use a member of the struct to provide the parameter offset instead.

Reply:
If I'm understanding the problem right, the way this is usually handled is to store function attributes at offset 0 and parameter attributes at offset i.

Reply:
I can think of 2 solutions to this problem. Eighter store function attributes like this:

```
0 => array( func attributes ... ),
1 => array ( param 1 attributes ),
3 => array (param 3 attributes )
```
This would require at wrapper array and may have index gaps. The other solution would be to store all function and parameter attributes in single list / array. The attribute struct would need to store the parameter offset. Finding parameter attributes would require traversing all the attributes and testing for the right offset. Since these lists should be pretty small it will likely be fast enough.